mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -261,58 +261,93 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
+  private[spark] def recommendForAll[T: ClassTag : Fractional](
       rank: Int,
-      srcFeatures: RDD[(Int, Array[Double])],
-      dstFeatures: RDD[(Int, Array[Double])],
-      num: Int): RDD[(Int, Array[(Int, Double)])] = {
-    val srcBlocks = blockify(rank, srcFeatures)
-    val dstBlocks = blockify(rank, dstFeatures)
+      srcFactors: RDD[(Int, Array[T])],
+      dstFactors: RDD[(Int, Array[T])],
+      num: Int): RDD[(Int, Array[(Int, T)])] = {
+    val srcBlocks = blockify(rank, srcFactors)
+    val dstBlocks = blockify(rank, dstFactors)
+    val tag = implicitly[ClassTag[T]].runtimeClass
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
       case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
         val m = srcIds.length
         val n = dstIds.length
-        val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
-        ratings.foreachActive { (i, j, r) =>
-          output(k) = (srcIds(i), (dstIds(j), r))
-          k += 1
+        val targetMatrix = new Array[T](m * n)
+        tag match {
+          case java.lang.Float.TYPE =>
+            val A = srcFactors.asInstanceOf[Array[Float]]
+            val B = dstFactors.asInstanceOf[Array[Float]]
+            val C = targetMatrix.asInstanceOf[Array[Float]]
+            blas.sgemm("T", "N", m, n, rank, 1f, A, rank, B, rank, 0f, C, m)
+          case java.lang.Double.TYPE =>
+            val A = srcFactors.asInstanceOf[Array[Double]]
+            val B = dstFactors.asInstanceOf[Array[Double]]
+            val C = targetMatrix.asInstanceOf[Array[Double]]
+            blas.dgemm("T", "N", m, n, rank, 1f, A, rank, B, rank, 0f, C, m)
         }
-        output.toSeq
+        fillPredictions(srcIds, dstIds, targetMatrix, m, n).toSeq
     }
     ratings.topByKey(num)(Ordering.by(_._2))
   }
 
   /**
+   * Fills array of (srcId, Array[(targetId, score)] from the result of multiplying the
+   * user and item factor matrices.
+   */
+  private def fillPredictions[T : ClassTag](
+      srcIds: Array[Int],
+      dstIds: Array[Int],
+      predictions: Array[T],
+      m: Int,
+      n: Int): Array[(Int, (Int, T))] = {
+    val output = new Array[(Int, (Int, T))](m * n)
+    // outer loop over columns
+    var k = 0
+    var j = 0
+    while (j < n) {
+      var i = 0
+      val indStart = j * m
+      while (i < m) {
+        output(k) = (srcIds(i), (dstIds(j), predictions(indStart + i)))
+        i += 1
+        k += 1
+      }
+      j += 1
+    }
+    output
+  }
+
+  /**
    * Blockifies features to use Level-3 BLAS.
    */
-  private def blockify(
+  private def blockify[T: ClassTag](
       rank: Int,
-      features: RDD[(Int, Array[Double])]): RDD[(Array[Int], DenseMatrix)] = {
-    val blockSize = 4096 // TODO: tune the block size
+      features: RDD[(Int, Array[T])],
+      /* TODO make blockSize a param */blockSize: Int = 4096): RDD[(Array[Int], Array[T])] = {
```

Comment:
L3 cache on recent Intel processors is 8MB, with a 64-byte cache line size. You'll want one block to fit into L3 cache to make dense matrix multiply perform well. Block size should be 1024 (1024*1024*8 bytes = 8MB) if the shared rank of the factors is > 1024; otherwise, block size should be 1048576 / (rank), rounded **down** to the nearest 64 bytes.